### PR TITLE
Allow rebuilding graphs when output type depends on input values

### LINKED ISSUE #253: `clone_replace` doesn't try to rebuild nodes when input types don't change
### Description

```python
import pytensor
import pytensor.tensor as pt

x = pt.tensor("x", shape=(None, 100))
y = pt.zeros(x.shape[-1])
new_y = pytensor.clone_replace(y, replace={x: pt.zeros((2, 50,))}, rebuild_strict=False)

print(x.type)  # TensorType(float64, (100,))
print(new_y.type)  # TensorType(float64, (100,)) <- INCORRECT!
print(new_y.eval().shape)  # (50,)
```

This happens because `clone_with_new_inputs` doesn't try to rebuild the node when the inputs have the same types as before: https://github.com/ricardoV94/pytensor/blob/8606498fa92e86ec10f7357e4da228cfd86bced7/pytensor/graph/basic.py#L269-L280

This is incorrect as the output type can change when the input values change, not just their types, as happens with `Alloc`.

I think we should always rebuild the node if `strict=False`

### LINKED ISSUE #254: Static type shapes break graph dependency
### Description

In the following example, calling `x.shape`, simply returns `TensorConstant(100)`, which means the `clone_replace` will have no effect.

```python
import pytensor
import pytensor.tensor as pt

x = pt.tensor("x", shape=(100,))
y = pt.zeros(x.shape[-1])
new_y = pytensor.clone_replace(y, replace={x: pt.zeros((50,))}, rebuild_strict=False)

print(x.shape)  # TensorConstant{(1,) of 100}
print(y.type)  # TensorType(float64, (100,))
print(new_y.type)  # TensorType(float64, (100,))
print(new_y.eval().shape)  # (100,)
```

Instead, if we add a single useless `None` dimension, `Shape` will return an object that is still linked to the original variable, and the replacement will works as intended (except for the issue discussed in #253).

```python
x = pt.tensor("x", shape=(None, 100,))
y = pt.zeros(x.shape[-1])
new_y = pytensor.clone_replace(y, replace={x: pt.zeros((2, 50,))}, rebuild_strict=False)

print(x.shape)  # Shape.0
print(y.type)  # TensorType(float64, (100,))
print(new_y.type)  # TensorType(float64, (100,))
print(new_y.eval().shape)  # (50,)
```

I suggest we don't "severe" the graph dependency between shape and original variables until the compilation stage. 
This will allow us to keep manipulating graphs, even as our static type inference continues to improve.

Otherwise we will lose much of PyTensor graph manipulation appeal.

The relevant logic is here:
https://github.com/pymc-devs/pytensor/blob/8606498fa92e86ec10f7357e4da228cfd86bced7/pytensor/tensor/shape.py#L144-L156

And in the quite redundant:
https://github.com/pymc-devs/pytensor/blob/8606498fa92e86ec10f7357e4da228cfd86bced7/pytensor/tensor/var.py#L261-L265

### PR DESCRIPTION
**Pull request by ricardoV94**
Closes #254 
Closes #253

### CODE REVIEW

File: tests/tensor/test_nlinalg.py [hunk 1]
```diff
@@ -274,9 +273,7 @@ def test_det_grad():
 
 def test_det_shape():
     x = matrix()
-    det_shape = det(x).shape
-    assert isinstance(det_shape, Constant)
-    assert tuple(det_shape.data) == ()
+    assert det(x).type.shape == ()
```
Comment (ferrine) [inline tests/tensor/test_nlinalg.py:276]: How does this check the intended behaviour?

File: tests/tensor/test_var.py [hunk 2]
```diff
@@ -245,8 +246,13 @@ def test__getitem__newaxis(x, indices, new_order):
 
 def test_fixed_shape_variable_basic():
     x = TensorVariable(TensorType("int64", shape=(4,)), None)
-    assert isinstance(x.shape, Constant)
-    assert np.array_equal(x.shape.data, (4,))
+    assert x.type.shape == (4,)
```
Comment (ferrine) [inline tests/tensor/test_var.py:249]: In the original issue you said that you should not get constants in data, can it be verified explicitly? So far it is not obvious for me

File: tests/tensor/test_var.py [hunk 3]
```diff
@@ -245,8 +246,13 @@ def test__getitem__newaxis(x, indices, new_order):
...
+
+    shape_fn = pytensor.function([x], x.shape)
+    opt_shape = shape_fn.maker.fgraph.outputs[0]
+    assert isinstance(opt_shape.owner.op, DeepCopyOp)
+    assert isinstance(opt_shape.owner.inputs[0], Constant)
```
Comment (ferrine) [inline tests/tensor/test_var.py:254]: This one is expected since you have compiled the graph, do I understand correctly?

Comment (ricardoV94) [reply tests/tensor/test_var.py:254]: Yes

**Comment by ricardoV94** [comment]
I test that the Shape Op is present even when static shapes are present.
I also added a second commit to fix #253

File: pytensor/tensor/basic.py [hunk 4]
```diff
@@ -1418,6 +1418,8 @@ class Alloc(COp):
     """
 
     _f16_ok = True
+    _output_type_depends_on_input_value = True
```
Comment (ferrine) [inline pytensor/tensor/basic.py:1421]: Why isn't it in a some base class with the default False value?

Comment (ricardoV94) [reply pytensor/tensor/basic.py:1421]: A lot of these flags are not in a base class like the `_f16_ok`, but no strong preference

Comment (ricardoV94) [reply pytensor/tensor/basic.py:1421]: Do you prefer that?

Comment (ricardoV94) [reply pytensor/tensor/basic.py:1421]: Added it to `Op`

File: pytensor/graph/basic.py [hunk 5]
```diff
@@ -266,8 +266,13 @@ def clone_with_new_inputs(
...
         remake_node = False
         new_inputs: List["Variable"] = list(inputs)
+
+        # Some Ops like Alloc require the node to always be rebuilt in non-strict mode
+        # as the output type depends on the input values and not just their types
+        output_type_depends_on_input_value = self.op._output_type_depends_on_input_value
+
         for i, (curr, new) in enumerate(zip(self.inputs, new_inputs)):
-            if curr.type != new.type:
+            if (curr.type != new.type) or output_type_depends_on_input_value:
```
Comment (ferrine) [inline pytensor/graph/basic.py:275]: Not super clear the branching logic from the outsider glance, can you please add comments that explain what exactly happens in each branch?

**Review by ferrine — CHANGES_REQUESTED** [review]

File: pytensor/graph/basic.py (hunk 5 above)
Comment (ricardoV94) [reply pytensor/graph/basic.py:276]: Done

**Review by ferrine — APPROVED** [review]

**Pull request merged**